flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java

```diff
@@ -0,0 +1,462 @@
+		this.notifyNewSlots = notifyNewSlots;
+		this.idleSlotTimeout = idleSlotTimeout;
+		this.rpcTimeout = rpcTimeout;
+		this.totalResourceRequirements = ResourceCounter.empty();
+		this.fulfilledResourceRequirements = ResourceCounter.empty();
+		this.slotToRequirementProfileMappings = new HashMap<>();
+	}
+
+	@Override
+	public void increaseResourceRequirementsBy(ResourceCounter increment) {
+		totalResourceRequirements = totalResourceRequirements.add(increment);
+
+		declareResourceRequirements();
+	}
+
+	@Override
+	public void decreaseResourceRequirementsBy(ResourceCounter decrement) {
+		totalResourceRequirements = totalResourceRequirements.subtract(decrement);
+
+		declareResourceRequirements();
+	}
+
+	private void declareResourceRequirements() {
+		notifyNewResourceRequirements.accept(getResourceRequirements());
+	}
+
+	@Override
+	public Collection<ResourceRequirement> getResourceRequirements() {
+		final Collection<ResourceRequirement> currentResourceRequirements = new ArrayList<>();
+
+		for (Map.Entry<ResourceProfile, Integer> resourceRequirement : totalResourceRequirements.getResourcesWithCount()) {
+			currentResourceRequirements.add(ResourceRequirement.create(resourceRequirement.getKey(), resourceRequirement.getValue()));
+		}
+
+		return currentResourceRequirements;
+	}
+
+	@Override
+	public Collection<SlotOffer> offerSlots(
+		Collection<? extends SlotOffer> offers,
+		TaskManagerLocation taskManagerLocation,
+		TaskManagerGateway taskManagerGateway,
+		long currentTime) {
+
+		LOG.debug("Received {} slot offers from TaskExecutor {}.", offers.size(), taskManagerLocation);
+		final Collection<SlotOffer> acceptedSlotOffers = new ArrayList<>();
+		final Collection<SlotOffer> candidates = new ArrayList<>();
+
+		// filter out already accepted offers
+		for (SlotOffer offer : offers) {
+			final AllocationID allocationId = offer.getAllocationId();
+			if (slotPool.containsSlot(allocationId)) {
+				acceptedSlotOffers.add(offer);
+			} else {
+				candidates.add(offer);
+			}
+		}
+
+		final Collection<SlotOfferMatching> matchings = matchOffersWithOutstandingRequirements(candidates);
+
+		final Collection<AllocatedSlot> acceptedSlots = new ArrayList<>();
+		ResourceCounter acceptedResources = ResourceCounter.empty();
+
+		for (SlotOfferMatching matching : matchings) {
+			if (matching.getMatching().isPresent()) {
+				final ResourceProfile matchedResourceProfile = matching.getMatching().get();
+
+				final AllocatedSlot allocatedSlot = createAllocatedSlot(
+					matching.getSlotOffer(),
+					taskManagerLocation,
+					taskManagerGateway);
+
+				acceptedSlots.add(allocatedSlot);
+				acceptedSlotOffers.add(matching.getSlotOffer());
+
+				acceptedResources = acceptedResources.add(matchedResourceProfile, 1);
+
+				// store the ResourceProfile against which the given slot has matched for future book-keeping
+				slotToRequirementProfileMappings.put(allocatedSlot.getAllocationId(), matchedResourceProfile);
```

Comment:
there are so many data-structures to modify (acceptedSlots/acceptedSlotOffers/acceptedResources), which would have to be passed in, along with other stuff (taskManagerLocation/Gateway), that it would become a bit of a mess.

Reply:
fair enough